README.md
# MCsniperGO
<a href="https://mcsniperpy.com/discord"><img src="https://invidget.switchblade.xyz/yp69ZqtxNk"/><a/>

This project was made possible by my [donators](https://kqzz.me/donators)

## Usage

> This sniper is in it's *beta* stage, meaning bugs should be expected.

### Easy installation

- download the latest release for your operating system in the [releases](https://github.com/Kqzz/MCsniperGO/releases/) tab
- move that file to it's own folder (recommended, issues may arise if not done)
- run that file once, 2 files will appear in that folder, `accounts.txt` and `config.toml`. `config.toml` can be ignored for now.
  - on windows, you can double click the executable file to run.
  - on macos, you must [open a terminal in the folder you moved the binary to](https://www.stugon.com/open-terminal-in-current-folder-location-mac/), run `chmod +x ./binary_name_here`, and then run`./binary_name_here` in the terminal
  - on linux, you must open a terminal and use `cd` to navigate to the folder where the binary is located, you may have to run `chmod +x ./binary_name_here`, and then run `./binary_name_here`
- open accounts.txt and add your accounts according to the formatting specified [below](https://github.com/Kqzz/MCsniperGO#accounts-file-format)
- run the sniper again with the same commands used before
  - for windows, you can double click the executable
  - for macos, you will have to open the terminal in the correct directory (as shown above) and run `./binary_name_here`
  - for linux, you will have to open a terminal and use `cd` to navigate to the correct directory, then run `./binary_name_here`
- the sniper will now prompt you for a username and offset. enter those and then the sniper will authenticate (this is run 8 hours before snipe) and then count down.

> A video guide will be made once this sniper is stabler

### Compiling from source (not recommended)

install [git](https://git-scm.com/) and [go](https://golang.org/dl/) 1.16 or later.
```sh
git clone https://github.com/Kqzz/MCsniperGO
cd MCsniperGO
go run .

```

## Accounts file format

You have to put accounts in accounts.txt before sniping, so follow the format shown below to add your accounts

> **in the code block shown below, replace any words fully capitalized with actual values, just leave the other words as is.**

```txt
# You can comment out lines by including a # at the start
# !! IMPORTANT !! ONLY REPLACE THE CAPITALIZED WORDS WITH THE CORRECT VALUE, EVERYTHING ELSE IS A "FLAG" FOR THE SNIPER TO KNOW WHAT TO DO WITH THE ACCOUNT.
# for example, for a prename snipe with manual bearer, do BEARERHERE:bearer:prename and DO NOT REPLACE "bearer" or "prename" with anything.

# MOJANG ACCOUNT

EMAIL:PASS
# OR (if you have security questions)
EMAIL:PASS:ANSWER:ANSWER:ANSWER

# Microsoft account
# THIS WILL PROMPT YOU TO LOGIN WHEN AUTHENTICATING, FOLLOW THE INSTRUCTIONS DISPLAYED WHEN YOU ARE PROMPTED.
# there is no password input. it will always be shown as 'oauth2-external' when printing a password.
IDENTIFYING_TEXT:ms
# OR FOR PRENAME
IDENTIFYING_TEXT:prename
# IT IS NOT RECOMMENDED TO USE THE EMAIL:PASS METHOD FOR THIS, SHOWN BELOW.
EMAIL:PASS:ms
# OR FOR PRENAME
EMAIL:PASS:prename

# Manual bearer
BEARERHERE:bearer

# You can add :prename for a prename snipe w/ manual bearer
BEARERHERE:bearer:prename
```


changes.md
I attempted to add a turbo command and also clean up some code / unnessasary things..

some things i suggest implementing or changing.

- your sniper function makes it hard to implement turbo and other commands i wanted to use. the setup itself is okay, but acc types and other info can easily be gained with a helper function that reqs the mojang api.
- because you dont cache bearers and reauth the accs every request, it makes it hard for improving the "system" for the way it was before it was doable, but if you want to add more commands your community wants it will get in the way some.

# Acc type function

```go
func isGC(bearer string) string {
	var accountT string
	conn, _ := tls.Dial("tcp", "api.minecraftservices.com"+":443", nil)

	fmt.Fprintln(conn, "GET /minecraft/profile/namechange HTTP/1.1\r\nHost: api.minecraftservices.com\r\nUser-Agent: Dismal/1.0\r\nAuthorization: Bearer "+bearer+"\r\n\r\n")

	e := make([]byte, 12)
	conn.Read(e)

	switch string(e[9:12]) {
	case `404`:
		accountT = "Giftcard"
	default:
		accountT = "Microsoft"
	}

	return accountT
}
```

# Bearer cache system

```go
func authAccs() {
	var AccountsVer []string
	file, _ := os.Open("accounts.txt")

	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		AccountsVer = append(AccountsVer, scanner.Text())
	}

	if len(AccountsVer) == 0 {
		sendE("Unable to continue, you have no accounts added.\n")
		os.Exit(0)
	}

	grabDetails(AccountsVer)

	if !acc.ManualBearer {
		if acc.Bearers == nil {
			sendE("No bearers have been found, please check your details.")
			os.Exit(0)
		} else {
			checkifValid()

			for _, acc := range acc.Bearers {
				if acc.NameChange {
					bearers.Details = append(bearers.Details, apiGO.Info{
						Bearer:      acc.Bearer,
						AccountType: acc.Type,
						Email:       acc.Email,
					})
				}
			}

			if bearers.Details == nil {
				sendE("Failed to authorize your bearers, please rerun the sniper.")
				os.Exit(0)
			}
		}
	}
}

func grabDetails(AccountsVer []string) {
	if acc.ManualBearer {
		for _, bearer := range AccountsVer {
			if apiGO.CheckChange(bearer) {
				bearers.Details = append(bearers.Details, apiGO.Info{
					Bearer:      bearer,
					AccountType: isGC(bearer),
				})
			}

			time.Sleep(time.Second)
		}
	} else {
		if acc.Bearers == nil {
			bearerz := apiGO.Auth(AccountsVer)
			if len(bearerz.Details) == 0 {
				sendE("Unable to authenticate your account(s), please Reverify your login details.\n")
				return
			} else {
				for _, accs := range bearerz.Details {
					acc.Bearers = append(acc.Bearers, apiGO.Bearers{
						Bearer:       accs.Bearer,
						AuthInterval: 86400,
						AuthedAt:     time.Now().Unix(),
						Type:         accs.AccountType,
						Email:        accs.Email,
						Password:     accs.Password,
						NameChange:   apiGO.CheckChange(accs.Bearer),
					})
				}
				acc.SaveConfig()
				acc.LoadState()
			}
		} else {
			if len(acc.Bearers) < len(AccountsVer) {
				var auth []string
				check := make(map[string]bool)

				for _, acc := range acc.Bearers {
					check[acc.Email+":"+acc.Password] = true
				}

				for _, accs := range AccountsVer {
					if !check[accs] {
						auth = append(auth, accs)
					}
				}

				bearerz := apiGO.Auth(auth)

				if len(bearerz.Details) != 0 {
					for _, accs := range bearerz.Details {
						acc.Bearers = append(acc.Bearers, apiGO.Bearers{
							Bearer:       accs.Bearer,
							AuthInterval: 86400,
							AuthedAt:     time.Now().Unix(),
							Type:         accs.AccountType,
							Email:        accs.Email,
							Password:     accs.Password,
							NameChange:   apiGO.CheckChange(accs.Bearer),
						})
					}

					acc.SaveConfig()
					acc.LoadState()
				}
			} else if len(AccountsVer) < len(acc.Bearers) {
				for _, accs := range AccountsVer {
					for _, num := range acc.Bearers {
						if accs == num.Email+":"+num.Password {
							acc.Bearers = append(acc.Bearers, num)
						}
					}
				}
				acc.SaveConfig()
				acc.LoadState()
			}
		}
	}
}

func checkifValid() {
	var reAuth []string
	for _, accs := range acc.Bearers {
		f, _ := http.NewRequest("GET", "https://api.minecraftservices.com/minecraft/profile/name/boom/available", nil)
		f.Header.Set("Authorization", "Bearer "+accs.Bearer)
		j, _ := http.DefaultClient.Do(f)

		if j.StatusCode == 401 {
			sendI(fmt.Sprintf("Account %v turned up invalid. Attempting to Reauth", accs.Email))
			reAuth = append(reAuth, accs.Email+":"+accs.Password)
		}
	}

	if len(reAuth) != 0 {
		sendI(fmt.Sprintf("Reauthing %v accounts..", len(reAuth)))
		bearerz := apiGO.Auth(reAuth)

		if len(bearerz.Details) != 0 {
			for point, data := range acc.Bearers {
				for _, accs := range bearerz.Details {
					if data.Email == accs.Email {
						data.Bearer = accs.Bearer
						data.NameChange = apiGO.CheckChange(accs.Bearer)
						data.Type = accs.AccountType
						data.Password = accs.Password
						data.Email = accs.Email
						data.AuthedAt = time.Now().Unix()
						acc.Bearers[point] = data
						acc.SaveConfig()
					}
				}
			}
		}
	}

	acc.LoadState()
}
```

With this you can replace my api with yours. aswell change the structs to your acc structs.

# Want a reauth system?..

```go
func checkAccs() {
	for {
		time.Sleep(time.Second * 10)

		// check if the last auth was more than a minute ago
		for _, accs := range acc.Bearers {
			if time.Now().Unix() > accs.AuthedAt+accs.AuthInterval {
				sendI(accs.Email + " is due for reauth")

				// authenticating account
				bearers := apiGO.Auth([]string{accs.Email + ":" + accs.Password})

				if bearers.Details != nil {
					for point, data := range acc.Bearers {
						for _, accs := range bearers.Details {
							if data.Email == accs.Email {
								data.Bearer = accs.Bearer
								data.NameChange = apiGO.CheckChange(accs.Bearer)
								data.Type = accs.AccountType
								data.Password = accs.Password
								data.Email = accs.Email
								data.AuthedAt = time.Now().Unix()
								acc.Bearers[point] = data
							}
						}

utils.go
package main

import (
	"bufio"
	"crypto/tls"
	"errors"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/ecnepsnai/discord"
	"github.com/kqzz/mcgo"
)

var errAccIgnored error = errors.New("account was ignored, either commented or otherwise")

// from gosnipe lmao
func pingMojang() (float64, error) {
	var sumNanos int64
	conn, err := tls.Dial("tcp", "api.minecraftservices.com:443", nil)
	if err != nil {
		return 0, err
	}

	defer conn.Close()
	for i := 0; i < 3; i++ {
		recv := make([]byte, 4096)
		time1 := time.Now()
		conn.Write([]byte("PUT /minecraft/profile/name/test HTTP/1.1\r\nHost: api.minecraftservices.com\r\nAuthorization: Bearer TestToken\r\n\r\n"))
		conn.Read(recv)
		sumNanos += time.Since(time1).Milliseconds()
	}

	sumNanos /= 3
	return float64(sumNanos / 1000000), nil
}

// readLines reads a whole file into memory
// and returns a slice of its lines.
func readLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func strSliceContainsMultiOption(s []string, strs []string) bool {
	for _, str := range strs {
		for _, v := range s {
			if v == str {
				return true
			}
		}
	}

	return false
}

func loadAccStr(accStr string) (mcgo.MCaccount, error) {
	if strings.HasPrefix(accStr, "#") {
		return mcgo.MCaccount{}, errAccIgnored
	}

	var account mcgo.MCaccount
	strSplit := strings.Split(accStr, ":")
	strSplitLower := strings.Split(strings.ToLower(accStr), ":")

	var AccType mcgo.AccType

	if strSliceContainsMultiOption(strSplitLower, []string{"ms", "msa"}) {
		AccType = mcgo.Ms
	} else if strSliceContainsMultiOption(strSplitLower, []string{"prename", "msprename", "msaprename", "pr"}) {
		AccType = mcgo.MsPr
	} else {
		AccType = mcgo.Mj
	}

	if strSliceContainsMultiOption(strSplitLower, []string{"bearer"}) {
		account = mcgo.MCaccount{
			Bearer: strSplit[0],
			Type:   AccType,
		}
		return account, nil
	}

	switch AccType {
	case mcgo.Mj:
		{
			switch len(strSplit) {
			case 2:
				{
					account = mcgo.MCaccount{
						Email:    strSplit[0],
						Password: strSplit[1],
						Type:     AccType,
					}
				}
			case 5:
				{
					account = mcgo.MCaccount{
						Email:           strSplit[0],
						Password:        strSplit[1],
						SecurityAnswers: strSplit[2:5],
						Type:            AccType,
					}
				}
			default:
				return account, fmt.Errorf("invalid split count of %v on line: %v", len(strSplit), accStr)
			}
		}
	case mcgo.Ms, mcgo.MsPr:
		{
			if len(strSplit) == 3 {
				account = mcgo.MCaccount{
					Email:    strSplit[0],
					Password: strSplit[1],
					Type:     AccType,
				}
			} else {
				account = mcgo.MCaccount{
					Email:    strSplit[0],
					Password: "oauth2-external",
					Type:     AccType,
				}
			}
		}
	}

	return account, nil
}

func loadAccSlice(accSlice []string) []*mcgo.MCaccount {
	var accounts []*mcgo.MCaccount
	for i, accStr := range accSlice {
		if accStr == "" {
			continue
		}
		acc, err := loadAccStr(accStr)
		if err != nil {
			if !errors.Is(err, errAccIgnored) {
				log("error", `got error "%v" while loading acc on line %v`, err, i+1)
			}
			continue
		}
		accounts = append(accounts, &acc)
	}
	return accounts
}

func genHeader() string {
	header := `
███╗   ███╗ ██████╗███████╗███╗   ██╗██╗██████╗ ███████╗██████╗  ██████╗  ██████╗ 
████╗ ████║██╔════╝██╔════╝████╗  ██║██║██╔══██╗██╔════╝██╔══██╗██╔════╝ ██╔═══██╗
██╔████╔██║██║     ███████╗██╔██╗ ██║██║██████╔╝█████╗  ██████╔╝██║  ███╗██║   ██║
██║╚██╔╝██║██║     ╚════██║██║╚██╗██║██║██╔═══╝ ██╔══╝  ██╔══██╗██║   ██║██║   ██║
██║ ╚═╝ ██║╚██████╗███████║██║ ╚████║██║██║     ███████╗██║  ██║╚██████╔╝╚██████╔╝
╚═╝     ╚═╝ ╚═════╝╚══════╝╚═╝  ╚═══╝╚═╝╚═╝     ╚══════╝╚═╝  ╚═╝ ╚═════╝  ╚═════╝ 
<fg=cyan>https://mcsniperpy.com</>

`

	for _, char := range []string{"╗", "║", "╝", "╔", "═"} {
		header = strings.ReplaceAll(header, char, fmt.Sprintf("<fg=white>%v</>", char))
	}

	header = strings.ReplaceAll(header, "█", "<fg=blue>█</>")

	return header
}

func fileExists(filename string) bool {
	_, err := os.Stat(filename)
	return !os.IsNotExist(err)
}

func prettyAccType(acc mcgo.AccType) string {
	return map[mcgo.AccType]string{
		mcgo.Mj:   "mojang",
		mcgo.Ms:   "microsoft",
		mcgo.MsPr: "microsoft prename",
	}[acc] + " account"
}

func countAccounts(accounts []*mcgo.MCaccount) (int, int) {
	normCount := 0
	prenameCount := 0
	for _, acc := range accounts {
		switch acc.Type {
		case mcgo.Mj, mcgo.Ms:
			{
				normCount += 1
			}
		case mcgo.MsPr:
			{
				prenameCount += 1
			}
		}
	}
	return normCount, prenameCount
}

func fmtTimestamp(timestamp time.Time) string {
	return timestamp.Format("05.999999")
}

func formatAccount(account *mcgo.MCaccount) string {
	return fmt.Sprintf("%v:%v | Type: %v", account.Email, account.Password, prettyAccType(account.Type))
}

func announceSnipe(username, auth string, account *mcgo.MCaccount) error {
	req, err := http.NewRequest("POST", fmt.Sprintf("https://api.mcsniperpy.com/announce?username=%v&prename=%v", username, account.Type == mcgo.MsPr), nil)

	if err != nil {
		return err
	}

	req.Header.Add("Authorization", auth)

	res, err := http.DefaultClient.Do(req)

	if err != nil {
		return err
	}

	defer res.Body.Close()

	switch res.StatusCode {
	case 401:
		{
			return errors.New("invalid auth code")
		}
	case 429:
		{
			return errors.New("too many requests (ask staff to manually announce)")
		}
	default:
		return errors.New(fmt.Sprintf("Got: %v Couldnt announce your snipe, please contact staff to manually post it.", res.StatusCode))
	}
}

func customServerAnnounce(name string) error {
	config, err := getConfig()
	if err != nil {
		panic(err)
	}
	discord.WebhookURL = config.Announce.WebhookURL

	discord.Post(discord.PostOptions{
		Embeds: []discord.Embed{
			{
				Footer: &discord.Footer{
					Text:    "MCsniperGO",
					IconURL: "https://cdn.discordapp.com/icons/734794891258757160/a_011d19e6e17a5eb46d108fd45b28dc9d.webp?size=96",
				},
				Title:       "Successful Snipe",
				URL:         "https://github.com/Kqzz/MCsniperGO",
				Color:       3118847,
				Description: fmt.Sprintf("Name: [`%v`](https://namemc.com/search?q=%v)", name, name),
			},
		},
	})
	return nil
}

func accID(acc *mcgo.MCaccount) string {
	if acc.Email != "" {
		return acc.Email
	} else if acc.Bearer != "" {
		return acc.Bearer[len(acc.Bearer)-10:]
	} else {
		return "<unknown account>"
	}
}

func authAccount(acc *mcgo.MCaccount) error {
	// authenticating if bearer isn't loaded
	if acc.Bearer == "" {
		switch acc.Type {
		case mcgo.MsPr, mcgo.Ms:
			{
				var err error
				if acc.Password == "oauth2-external" {
					err = acc.InitAuthFlow()
				} else {
					err = acc.MicrosoftAuthenticate()
				}
				if err != nil {
					return err
				}
				log("info", "authenticating %s through ms auth", accID(acc))
			}
		case mcgo.Mj:
			{
				err := acc.MojangAuthenticate()
				if err != nil {
					return err
				}
				log("info", "authenticating %s through mojang auth", accID(acc))
			}
		}
	} else {
		log("info", "authing %s through manual bearer", accID(acc))
	}

	return nil
}

func accReadyToSnipe(acc *mcgo.MCaccount) (bool, error) {
	switch acc.Type {
	case mcgo.MsPr:
		{
			canCreateProfile, err := acc.HasGcApplied()

			if err != nil {
				return false, fmt.Errorf("prename acc: %v", err)
			}

			return canCreateProfile, nil
		}
	case mcgo.Mj, mcgo.Ms:
		{
			nameChangeInfo, err := acc.NameChangeInfo()

			if err != nil {
				return false, fmt.Errorf("mj / ms acc: %v", err)
			}

			return nameChangeInfo.Namechangeallowed, nil

		}
	}
	return false, nil
}

func prettyStatus(status int) string {
	color := "red"
	if status < 300 {
		color = "green"
	}
	return fmt.Sprintf("<fg=%v;op=underscore>%v</>", color, status)
}
